### PR TITLE
Device objects

### PR DESCRIPTION
**Pull request by tommy-waltmann**
[ci skip]

## Description

I have implemented the device api for the transition to 3.0. This involved removing the global Execution Configruation, MPI Configuration, and Messenger object. Execution and Messenger objects are now encapsulted by the device, and the MPI configuration is encapsulated by the new Communicator object, which is held by a device. A device is held by a SimulationContext.

Some global functions in comm.py were moved to be properties or instance methods of the communicator.

The command line options have largely been stripped. Those options which were stripped can be set in the api of the device or the communicator objects.

## Motivation and Context

We are moving towards creating hoomd 3.0 which will not have global state as in previous versions. In order to achieve this, we needed to begin the process of removing many of the global variables that exist in the current version. Creating the device and communicator api is a large step towards removing much of the global state that currently exists in hoomd.

Resolves: #395 (although more work still needs to be done)

## How Has This Been Tested?

All the current unit tests were updated to reflect the changes that were made to the api. In addition, a new unit test module test_device.py was created in the hoomd/test-py directory, which tests the api for the device objects.

## Change log

```
New Features
    HOOMD
        Device api is now used instead of command line options
```

## Checklist:

- [x ] I have reviewed the [**Contributor Guidelines**](https://github.com/glotzerlab/hoomd-blue/blob/master/CONTRIBUTING.md).
- [ x] I agree with the terms of the [**HOOMD-blue Contributor Agreement**](https://github.com/glotzerlab/hoomd-blue/blob/master/ContributorAgreement.md).
- [ x] My name is on the [list of contributors](https://github.com/glotzerlab/hoomd-blue/blob/master/sphinx-doc/credits.rst).
- [ x] I am a member of the [hoomd-contributors team](https://github.com/orgs/glotzerlab/teams/hoomd-contributors/members).

## One Final Note:

Github still gives that red x with regard to unit tests on the device-objects branch, but that is because I accidentally started the unit tests a second time on my last push, even though it passed all the unit tests the first time. I quickly killed the second round of unit tests in order to not waste resources, and github thinks there is something wrong.

### CODE REVIEW

File: hoomd/device.py [hunk 1]
```diff
@@ -0,0 +1,381 @@
...
+    return msg
+
+
+def _init_nthreads(nthreads):
+    """
+    initializes the number of threads
+    """
+    
+    if nthreads is not None:
+        self.num_threads = nthreads
```
Comment (joaander) [inline hoomd/device.py]: Is it valid to pass `None` here? `None` is the default argument for `nthreads` in `CPU` and `Auto` below.

**Comment by joaander** [comment]
Reviewers, take a moment and offer a few comments on this PR and approve or request changes. I don't expect all reviewers to be experts at all portions of a PR. Multiple reviews from different perspectives provide valuable feedback to the project.

The most important changes are in `hoomd/device.py` and `hoomd/comm.py`. Preview the documentation for these modules:
- https://hoomd-blue.readthedocs.io/en/device-objects/module-hoomd-device.html
- https://hoomd-blue.readthedocs.io/en/device-objects/module-hoomd-comm.html

File: hoomd/context.py [hunk 2]
```diff
@@ -178,14 +179,12 @@ def __exit__(self, exc_type, exc_value, traceback):
 
         current = self.prev;
 
-def initialize(args=None, memory_traceback=False, mpi_comm=None):
+def initialize(args=None, device=None):
```
[the remark this answers is not in the document]
Comment (tommy-waltmann) [reply hoomd/context.py]: There are some unit tests which still rely on this. It should be removed later in the 3.0 transition when we re-write the unit tests.

File: hoomd/device.py [hunk 3]
```diff
@@ -0,0 +1,381 @@
...
+    username (str): the username
+    wallclocktime (float): elapsed time since the hoomd script first started execution
+    cputime (float): CPU clock time elapsed since the hoomd script first started execution
+    job_id (int): the id of the job
+    job_name (str): the job name
+    notice_level (int): minimum level of notice messages to print, settable
+    hostname (str): the network hostname
+    memory_tracback (bool): If true, enable memory allocation tracking (*only for debugging/profiling purposes*)
+    gpu_error_checking (bool): Whether or not CUDA error checking is enabled, settable.
+"""
```
[the remark this answers is not in the document]
Comment (vyasr) [reply hoomd/device.py]: I think the goal for metadata is to capture complete information up to restartability. That means it needs all information that would be necessary for another user to reproduce the run. From that perspective, I think it should be safe to remove the properties that you listed.

**Comment by jglaser** [comment]
How are multiple communicators handled? I see that you can now pass in a hoomd.comm.Communicator() object to the device, which is a good idea. mpi4py would allow creating multiple instances of those through splitting. What happens if you try to pass the same Communicator to multiple devices, or instantiate multiple Communicators in the same MPI runtime environment?

**Pull request merged**